services/src/test/java/io/grpc/services/HealthCheckingLoadBalancerFactoryTest.java

```diff
@@ -0,0 +1,1102 @@
+    // But the backoff policy has been reset.  A new backoff policy will be used for
+    // the next backed-off retry.
+    inOrder.verify(origLb).handleSubchannelState(
+        same(subchannel), eq(ConnectivityStateInfo.forNonError(CONNECTING)));
+    assertThat(healthImpl.calls).hasSize(1);
+    assertThat(clock.getPendingTasks()).isEmpty();
+    inOrder.verifyNoMoreInteractions();
+
+    // then closes the stream for this retry
+    healthImpl.calls.poll().responseObserver.onError(Status.UNAVAILABLE.asException());
+    inOrder.verify(origLb).handleSubchannelState(
+        same(subchannel),
+        unavailableStateWithMsg(
+            "Health-check stream was erroneously closed with " + Status.UNAVAILABLE
+            + " for 'TeeService'"));
+
+    // New backoff policy is used
+    inOrder.verify(backoffPolicyProvider).get();
+    // Retry with a new backoff policy
+    inOrder.verify(backoffPolicy2).nextBackoffNanos();
+
+    verifyRetryAfterNanos(inOrder, subchannel, healthImpl, 12);
+  }
+
+  private void verifyRetryAfterNanos(
+      InOrder inOrder, Subchannel subchannel, HealthImpl impl, long nanos) {
+    assertThat(impl.calls).isEmpty();
+    clock.forwardNanos(nanos - 1);
+    assertThat(impl.calls).isEmpty();
+    inOrder.verifyNoMoreInteractions();
+    verifyNoMoreInteractions(origLb);
+    clock.forwardNanos(1);
+    inOrder.verify(origLb).handleSubchannelState(
+        same(subchannel), eq(ConnectivityStateInfo.forNonError(CONNECTING)));
+    assertThat(impl.calls).hasSize(1);
+  }
+
+  @Test
+  public void serviceConfigHasNoHealthCheckingInitiallyButDoesLater() {
+    // No service config, thus no health check.
+    hcLbEventDelivery.handleResolvedAddressGroups(resolvedAddressList, Attributes.EMPTY);
+
+    verify(origLb).handleResolvedAddressGroups(same(resolvedAddressList), same(Attributes.EMPTY));
+    verifyNoMoreInteractions(origLb);
+
+    // First, create Subchannels 0
+    createSubchannel(0, Attributes.EMPTY);
+
+    // No health check activity.  Underlying Subchannel states are directly propagated
+    hcLbEventDelivery.handleSubchannelState(
+        subchannels[0], ConnectivityStateInfo.forNonError(READY));
+    assertThat(healthImpls[0].calls).isEmpty();
+    verify(origLb).handleSubchannelState(
+        same(subchannels[0]), eq(ConnectivityStateInfo.forNonError(READY)));
+
+    verifyNoMoreInteractions(origLb);
+
+    // Service config enables health check
+    Attributes resolutionAttrs = attrsWithHealthCheckService("FooService");
+    hcLbEventDelivery.handleResolvedAddressGroups(resolvedAddressList, resolutionAttrs);
+    verify(origLb).handleResolvedAddressGroups(
+        same(resolvedAddressList), same(resolutionAttrs));
+
+    // Health check started on existing Subchannel
+    assertThat(healthImpls[0].calls).hasSize(1);
+    verify(origLb).handleSubchannelState(
+        same(subchannels[0]), eq(ConnectivityStateInfo.forNonError(CONNECTING)));
```

Comment:
Going into CONNECTING at this point implies that if we push a service config that enables health checking, it will cause all new RPCs to be delayed until we get back the first health-check response from the server, which seems a little sub-optimal.  How hard would it be to special-case this such that we stay READY until the initial health-check request attempt completes?

Note that in C-core, we handle this by virtue of the fact that the service config is part of the identify of the subchannel in the global subchannel index.  This means that when we get a new service config enabling health checks, we create a new set of subchannels, so we handle it like any other resolver update: we keep using the old set of subchannels until the first of the new set of subchannels becomes READY, at which point we swap out the old list for the new list.  Of course, this approach does have its down-sides; most notably, the fact that we wind up with a lot of connection churn whenever the service config changes, which we're willing to live with (at least for now), because service config changes should be relatively rare.

If this is not something we can easily fix, then I think we can live with the behavior you have here, since the same logic applies: service config changes are fairly rare, so this shouldn't be a big deal.  But if we can easily do something smarter here, it would be good to do so.

Reply:
Done. I defined the special case as "If current state is READY, don't transition to CONNECTING when starting health check RPC".